pylzy/lzy/env/container/base.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from enum import Enum
+from typing import Optional, ClassVar, TYPE_CHECKING
+from typing_extensions import Protocol, runtime_checkable
+
+from lzy.env.base import Deconstructible
+
+if TYPE_CHECKING:
+    from lzy.env.mixin import WithEnvironmentType
+
+
+class DockerPullPolicy(Enum):
+    ALWAYS = "ALWAYS"  # Always pull the newest version of image
+    IF_NOT_EXISTS = "IF_NOT_EXISTS"  # Pull image once and cache it for next executions
+
+
+@runtime_checkable
+class DockerProtocol(Protocol):
+    @abstractmethod
+    def get_image_url(self) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_pull_policy(self) -> DockerPullPolicy:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_username(self) -> Optional[str]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_password(self) -> Optional[str]:
+        raise NotImplementedError
+
+
+@runtime_checkable
+class NoContainer(Protocol):
```

Comment:
Every class matches this protocol

Reply:
Yeah, but also class have to have "container_type = NoContainer" and be descend of BaseContainer to fulfill all my demands.